presto-main/src/main/java/com/facebook/presto/operator/LongArrayBlockEncodingBuffers.java

```diff
@@ -0,0 +1,188 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator;
+
+import io.airlift.slice.SliceOutput;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+
+import static com.facebook.presto.operator.ByteArrayUtils.writeLengthPrefixedString;
+import static com.google.common.base.Verify.verify;
+import static io.airlift.slice.SizeOf.SIZE_OF_INT;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.Math.max;
+import static sun.misc.Unsafe.ARRAY_LONG_INDEX_SCALE;
+
+public class LongArrayBlockEncodingBuffers
+        extends BlockEncodingBuffers
+{
+    private static final String NAME = "LONG_ARRAY";
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(LongArrayBlockEncodingBuffers.class).instanceSize();
+
+    private byte[] valuesBuffer;
+    private int valuesBufferIndex;
+
+    public LongArrayBlockEncodingBuffers(int initialBufferPositions)
+    {
+        super(initialBufferPositions);
+        prepareBuffers();
+    }
+
+    @Override
+    protected void prepareBuffers()
+    {
+        valuesBuffer = new byte[initialBufferPositions * ARRAY_LONG_INDEX_SCALE];
+        valuesBufferIndex = 0;
+    }
+
+    @Override
+    public void resetBuffers()
+    {
+        bufferedPositionCount = 0;
+        valuesBufferIndex = 0;
+        resetNullsBuffer();
+    }
+
+    @Override
+    public void accumulateRowSizes(int[] rowSizes)
+    {
+        int averageElementSize = Long.BYTES + Byte.BYTES;
+
+        for (int i = 0; i < positionCount; i++) {
+            rowSizes[i] += averageElementSize;
+        }
+    }
+
+    @Override
+    protected void accumulateRowSizes(int[] positionOffsets, int positionCount, int[] rowSizes)
+    {
+        int averageElementSize = Long.BYTES + Byte.BYTES;
+
+        // it starts from 0 because top level offsets were normalized
+        int lastOffset = 0;
+        for (int i = 0; i < positionCount; i++) {
+            int currentOffset = positionOffsets[i];
+            int entryCount = currentOffset - lastOffset;
+            rowSizes[i] += entryCount * averageElementSize;
+
+            lastOffset = currentOffset;
+        }
+    }
+
+    @Override
+    public void copyValues()
+    {
+        if (batchSize == 0) {
+            return;
+        }
+
+        verify(decodedBlock != null);
+
+        int[] positions = getPositions();
+        verify(positionsOffset + batchSize - 1 < positions.length && positions[positionsOffset] >= 0 &&
```

Comment:
I'd remove this check. If you decide to keep it, consider separating conjuncts into separate calls and moving these into `appendValuesToBuffer`.

```
        verify(positionsOffset + batchSize - 1 < positions.length);
        verify(positions[positionsOffset] >= 0);
        verify(positions[positionsOffset + batchSize - 1] < decodedBlock.getPositionCount());
```